### ISSUE TITLE
add 'ceammc/objectname' as an alternative creator for the object

### ISSUE THREAD

**Issue by porres** [uliss/pure-data#99]
we do that in cyclone so one can load cyclone/>~ instead of another and avoid conflicts.

for instance, now, if I have ceammc installed in vanilla and use another external that has the same name, such as else/xfade~, I can run into conflict problems. In the ideal world, one should be able to specify both objects and have control over which one is loaded where. The only way to do this with ceammc is giving it this option that is kind of a hack, I know

**Comment by uliss** [comment]
Did you added  this prefix to all "global" names like: **[>~]**, **[xfade~]** etc.?

**Comment by porres** [comment]
Hi, >~ is from cyclone, xfade~ is from ELSE. In cyclone, there's a library that loads only 12 objects (including >~), so I only added it for them, see https://github.com/porres/pd-cyclone/blob/master/cyclone_objects/binaries/cyclone_lib.c

the rest doesn't need this because they already work with the prefix, for instance: cyclone/comb~ else/xfade~ ect...

the problem here is that if I'm using ELSE and ceammc and I load else/xfade~ first, now there's no way to load xfade~ from ceammc anymore... the other way around has no problem. If I first load xfade~ from ceammc, I can use else/xfade~ and call the other one instead. But then, from then on, ceammc's xfade~ can't be recalled...

I guess one other solution is to use more complicated names that are just rarer, and you already do that for many objects, because you have your system that separates the objects in groups, such as synth.dx7~ alongside other "synth" objects.

So if you have a group for xfade~ and other objects that are not grouped and can have potential conflicts, that could also work without the need to add a classe creator with ceammc/objectname. But you should keep just "xfade~" also working as an alias for backwards compatibility anyway.

By the way, another object that has a conflict in your library is "fluid~" as there's a very old external out there that has the same name. That was one of the reasons for me not to upload the old fluid~ to deken, to avoid confusion, and I'm actually working on a new version of mne called "fluidsynth~" to avoid clashes. If not by adding "ceammc/" as a creator, maybe using synth.fluid~ and putting it also in the synth group works.

**Comment by smomid** [comment]
Hi, I just wanted to chime in on this issue as it's affecting me as well. I installed ceammc then I started working with some of my old patches. I noticed all my [gate] objects were not working as expected because they were over ridden by something in ceammc. Also all my [gain~] objects, which I think is in conflict with something from else~ weren't working. This kind of made my patch unusable because I use a lot of [gate]s all over the place and it would take awhile to replace all of them. 

Is there a way to come up with unique identifiers or names so there is no conflicts with objects from other libraries? I don't know which libraries are older, so please don't take offense at my comment, because I just discovered ceammc and I would like to use it, but there's so many conflicts with other things that I can't. 

thanks. Nick

**Comment by porres** [comment]
> I don't know which libraries are older

I don't think that matters, Pd should just do a better job allowing us to manage this and, unfortunately, we have this weak point in respect to compiled libraries. I brought this to the Pd's issue list as well, for reference https://github.com/pure-data/pure-data/issues/1257

**Comment by smomid** [comment]
if there are any workarounds for using these libraries at the same time please update me. Thanks for the info.

**Comment by porres** [comment]
In Pd, you should use [declare] for your patches to have better control of what you want t load.

You can easily force Pd to load a separate binary or abstraction even though you have something like ceammc loaded that has an object with the same name. For that, use prefixes like "else/xfade~"

The only real problem is if you want to use both objects with the same name. Once you load "else/xfade~", for instance, you can't load the one from ceammc.

**Comment by porres** [comment]
I was wrong so I deleted my last post. Things are pretty complicated and confusing...

Well, in Pd, you should use [declare] for your patches to have better control of what you want to load. In Pd-ceammc it's like it loads ceammc via startup and that makes a difference as I just realized.

You can easily force Pd to load a separate binary or abstraction even though you have something like ceammc (loaded via [declare -lib]) that has an object with the same name. For that, use prefixes like "else/xfade~". So, in Pd, with [declare], you can use both objects and use the one from ceammc if you don't use the prefix.

The only real problem is if you want to use both objects with the same name and you have ceammc loaded in Pd via startup, which is like pd-ceammc works. This way, once you load "else/xfade~", you can't load the one from ceammc anymore.

**Comment by porres** [comment]
new issue in pd's github :) https://github.com/pure-data/pure-data/issues/1271

**Comment by smomid** [comment]
Porres thanks for the tip. As a test I put ceammc in a folder outside of my seachpath. i made an abstraction with [declate -path /sequences/ceammc] in it, which shows where the folder is. I put a few objects from ceammc in the abstraction. When I quit pd and restarted, and opened the abstraction those ceammc objects didn't load. I tried a few different versions of declare like [declate -path ./sequences/ceammc]  and [declate -path ../sequences/ceammc] 

it's not that big of a deal, but just letting you know that method wasn't working for me, unless I did something wrong. thanks for looking into it.

**Comment by porres** [comment]
> unless I did something wrong

ceammc loads as a lib, so you need [declare -lib ceammc]

**Comment by smomid** [comment]
OK here's the problem. I created an abstraction with [declare -lib ceammc] inside an abstraction. And it loaded ceammc for all of my Pd patch including outside the abstraction, so I am again having conflicts with other objects. I thought if I declared a -path to ceammc that was outside of my search path for Pd, inside an abstraction,  I could use some of those objects inside the  abstraction and have everything outside the abstraction not have the conflict of the names. I hope that makes sense.
